### PR TITLE
fix: navigation for exisiting report screen

### PR DESCRIPTION
**Pull request by parasharrajat**
### Fixed Issues

$ https://github.com/Expensify/App/issues/4545

### Tests | QA Steps
1. Open New Dot on Mobile.
2. Go to Settings -> About.
3. Click on `Report a bug`.
4. Concierge chat will be opened now.
5. Now repeat steps 1 & 2.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web | Desktop

https://user-images.githubusercontent.com/24370807/132495294-5fb5021b-61b7-48f8-9535-7c1f74157cd6.mp4

#### Mobile Web

https://user-images.githubusercontent.com/24370807/132495014-80961876-7c27-4e9e-a605-daab7cb3319b.mp4

#### Android

https://user-images.githubusercontent.com/24370807/132494957-bc1d84a5-170d-4547-a325-0bb15151c5d1.mp4

### CODE REVIEW

File: src/libs/Navigation/CustomActions.js [hunk 1]
```diff
@@ -24,7 +24,19 @@ function pushDrawerRoute(screenName, params, navigationRef) {
 
         if (activeReportID === params.reportID) {
             if (state.type !== 'drawer') {
-                navigationRef.current.dispatch(StackActions.pop());
+                navigationRef.current.dispatch(() => {
+                    // If there are multiple routes then we can pop back to the first route
```
Comment (marcaaron) [inline src/libs/Navigation/CustomActions.js:28]: This code looks familiar... 😄 

Are you OK with just copying and pasting things from other places? 
Or should we think of a better way to handle this duplication?

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: src/libs/Navigation/CustomActions.js (hunk 1 above)
Comment (parasharrajat) [reply src/libs/Navigation/CustomActions.js:28]: Sorry about that. I just found this to be straight to start with this. I can definitely refactor it into a function. :smile:

Comment (parasharrajat) [reply src/libs/Navigation/CustomActions.js:28]: I am not sure the best way to dry this up. I can basically call the `DismissModal` here instead. It does basically the same thing. What do you suggest?

Comment (marcaaron) [reply src/libs/Navigation/CustomActions.js:28]: Please take some time to look into it and propose some options?

Comment (parasharrajat) [reply src/libs/Navigation/CustomActions.js:28]: Sure. Thanks.

File: src/libs/Navigation/Navigation.js [hunk 2]
```diff
@@ -114,39 +114,10 @@ function dismissModal(shouldOpenDrawer = false) {
...
     }
-
-    // Navigate back to where we were before we launched the modal
-    goBack(shouldOpenDrawer);
-
-    if (!normalizedShouldOpenDrawer) {
-        return;
-    }
-
-    openDrawer();
```
Comment (parasharrajat) [inline src/libs/Navigation/Navigation.js:149]: @marcaaron I didn't understand this part. Why were this code opening Drawer two times?
1. via goback when `shouldOpenDrawer` is `true`.
2. Then OpenDrawer call which happens when `shouldOpenDrawer` is `true`.

**Review by parasharrajat — COMMENTED** [review]
Code Updated. Please review.

File: src/libs/Navigation/Navigation.js (hunk 2 above)
Comment (marcaaron) [reply src/libs/Navigation/Navigation.js:149]: Can you rephrase this question? I'm not really sure what you are asking.

Comment (parasharrajat) [reply src/libs/Navigation/Navigation.js:149]: I looked at the code. It seems to me that this piece of code was calling `openDrawer` two times.

**Comment by marcaaron** [comment]
Just a heads up this is very far down my list of priorities and will take some time to review. 
Most likely I will be able to get to it within 2 weeks.

**Comment by kadiealexander** [comment]
Hi @parasharrajat, I've placed this issue on hold as per [this update](https://expensify.slack.com/archives/C01GTK53T8Q/p1631754482374800), as we are prioritising issues related to a feature release scheduled for 9/24. As an apology for the delay, we will add a $100 bonus as a thank you for waiting.

**Comment by parasharrajat** [comment]
@marcaaron Ready for review.

File: src/libs/Navigation/CustomActions.js [hunk 3]
```diff
@@ -1,6 +1,42 @@
 import {CommonActions, StackActions, DrawerActions} from '@react-navigation/native';
 import lodashGet from 'lodash/get';
 
+/**
+ * Go back to the Main Drawer
+ * @param {Object} navigationRef
+ */
+function navigateBackToDrawer(navigationRef) {
```
Comment (marcaaron) [inline src/libs/Navigation/CustomActions.js:8]: I'm having trouble with this name because it implies we are calling this when the drawer is not active. It looks like something I would call if I was not on the drawer to take me back to the drawer. However, if you look we also have logic for handling when the drawer _is_ active.

Ultimately, I find handling both responsibilities confusing and would love to figure out an alternative.

File: src/libs/Navigation/CustomActions.js [hunk 4]
```diff
@@ -24,7 +60,7 @@ function pushDrawerRoute(screenName, params, navigationRef) {
 
         if (activeReportID === params.reportID) {
             if (state.type !== 'drawer') {
-                navigationRef.current.dispatch(StackActions.pop());
+                navigateBackToDrawer(navigationRef);
```
Comment (marcaaron) [inline src/libs/Navigation/CustomActions.js:63]: One thing that sticks out to me is that we are checking `state.type !== 'drawer'` and then again checking in `navigateBackToDrawer()`. So it seems the method has more responsibility than it needs.

File: src/libs/Navigation/CustomActions.js [hunk 5]
```diff
@@ -1,6 +1,42 @@
...
+    let isLeavingDrawerNavigator = false;
+
+    // This should take us to the first view of the modal's stack navigator
+    navigationRef.current.dispatch((state) => {
+        // If this is a nested drawer navigator then we pop the screen and
+        // prevent calling goBack() as it's default behavior is to toggle open the active drawer
+        if (state.type === 'drawer') {
+            isLeavingDrawerNavigator = true;
+            return StackActions.pop();
+        }
```
Comment (marcaaron) [inline src/libs/Navigation/CustomActions.js:18]: Do we still have nested drawer navigators in use? I think we got rid of the one case. In any case, I think if we want to keep this logic we should make a distinction between a "nested drawer navigator" and a "main" or "root" drawer navigator. 

Otherwise, these variable names are confusing. I understand this code but someone might have a thought like, "Why do we have `isLeavingDrawerNavigator` in a method called `navigateBackToDrawer()`? Shouldn't `isLeavingDrawerNavigator` always be `false`?" Maybe it should be `isLeavingNestedDrawerNavigator` instead.

**Review by marcaaron — CHANGES_REQUESTED** [review]
Really like the clean up here and eliminating the need to explicitly call `Navigation.dismissModal()` in some places. But I have some thoughts and suggestions about how to make the code clearer.

**Review by marcaaron — APPROVED** [review]
Thanks for the updates 👍

File: src/libs/Navigation/CustomActions.js [hunk 6]
```diff
@@ -52,4 +88,5 @@ function pushDrawerRoute(screenName, params, navigationRef) {
 
 export default {
     pushDrawerRoute,
+    navigateBackToDrawer: navigateBackToRootDrawer,
```
Comment (parasharrajat) [inline src/libs/Navigation/CustomActions.js:91]: @marcaaron Oh, my editor messed up. Should I send another PR now?

Comment (parasharrajat) [reply src/libs/Navigation/CustomActions.js:91]: Follow up https://github.com/Expensify/App/pull/5937
